src/backend/cpu/triangle.cpp

```diff
@@ -0,0 +1,68 @@
+/*******************************************************
+ * Copyright (c) 2014, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#include <af/dim4.hpp>
+#include <af/defines.h>
+#include <Array.hpp>
+#include <triangle.hpp>
+#include <math.hpp>
+
+namespace cpu
+{
+
+template<typename T, bool is_upper>
```

Comment:
Yes, but everything down stream from this function will be compiled as two functions. is_upper needs to be a template inside the kernel for the best performance.